File: src/test/java/at/ac/uibk/dps/cirrina/execution/service/ServiceImplementationSelectorTest.java
package at.ac.uibk.dps.cirrina.execution.service;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;

import at.ac.uibk.dps.cirrina.csm.ServiceImplementationBindings.HttpMethod;
import at.ac.uibk.dps.cirrina.csm.ServiceImplementationBindings.HttpServiceImplementationBinding;
import at.ac.uibk.dps.cirrina.csm.ServiceImplementationBindings.ServiceImplementationBinding;
import at.ac.uibk.dps.cirrina.csm.ServiceImplementationBindings.Type;
import java.util.List;
import org.junit.jupiter.api.Test;

class ServiceImplementationSelectorTest {

  @Test
  void testSelectMatchingServices() {
    final var serviceBindings = new ServiceImplementationBinding[5];

    // Service one
    {
      final var service = new HttpServiceImplementationBinding(
        "A",
        true,
        Type.HTTP,
        "http",
        "localhost",
        12345,
        "",
        HttpMethod.GET
      );

      serviceBindings[0] = service;
    }

    // Service two
    {
      final var service = new HttpServiceImplementationBinding(
        "A",
        false,
        Type.HTTP,
        "http",
        "localhost",
        12345,
        "",
        HttpMethod.GET
      );

      serviceBindings[1] = service;
    }

    // Service three
    {
      final var service = new HttpServiceImplementationBinding(
        "B",
        false,
        Type.HTTP,
        "http",
        "localhost",
        12345,
        "",
        HttpMethod.GET
      );

      serviceBindings[2] = service;
    }

    // Service four
    {
      final var service = new HttpServiceImplementationBinding(
        "B",
        false,
        Type.HTTP,
        "http",
        "localhost",
        12345,
        "",
        HttpMethod.GET
      );

      serviceBindings[3] = service;
    }

    // Service five
    {
      final var service = new HttpServiceImplementationBinding(
        "C",
        true,
        Type.HTTP,
        "http",
        "localhost",
        12345,
        "",
        HttpMethod.GET
      );

      serviceBindings[4] = service;
    }

    final var services = ServiceImplementationBuilder.from(List.of(serviceBindings)).build();

    final var serviceSelector = new RandomServiceImplementationSelector(services);

    assertDoesNotThrow(() -> {
      var selected = serviceSelector.select("A", false);
    });

    // TODO: Add additional tests
  }
}

File: src/test/kotlin/at/ac/uibk/dps/cirrina/execution/service/ServiceImplementationSelectorTest.kt
package at.ac.uibk.dps.cirrina.execution.service

import at.ac.uibk.dps.cirrina.csm.Csml.HttpMethod
import at.ac.uibk.dps.cirrina.csm.Csml.HttpServiceImplementationBinding
import at.ac.uibk.dps.cirrina.csm.Csml.InvocationMode
import at.ac.uibk.dps.cirrina.csm.Csml.Type
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow

class ServiceImplementationSelectorTest {

  @Test
  fun testSelectMatchingServices() {
    RandomServiceImplementationSelector(
        ServiceImplementation.from(
            listOf(
              HttpServiceImplementationBinding(
                "A",
                true,
                Type.HTTP,
                "http",
                "localhost",
                12345,
                "",
                HttpMethod.GET,
              ),
              HttpServiceImplementationBinding(
                "A",
                false,
                Type.HTTP,
                "http",
                "localhost",
                12345,
                "",
                HttpMethod.GET,
              ),
              HttpServiceImplementationBinding(
                "B",
                false,
                Type.HTTP,
                "http",
                "localhost",
                12345,
                "",
                HttpMethod.GET,
              ),
              HttpServiceImplementationBinding(
                "B",
                false,
                Type.HTTP,
                "http",
                "localhost",
                12345,
                "",
                HttpMethod.GET,
              ),
              HttpServiceImplementationBinding(
                "C",
                true,
                Type.HTTP,
                "http",
                "localhost",
                12345,
                "",
                HttpMethod.GET,
              ),
            )
          )
          .getOrThrow()
      )
      .let { serviceImplementationSelector ->
        assertDoesNotThrow {
          // Success case
          assertNotNull(
            serviceImplementationSelector.select("A", InvocationMode.REMOTE),
            "selector should find a remote implementation for service 'A'",
          )

          // Missing service case
          assertNull(serviceImplementationSelector.select("D", InvocationMode.REMOTE))
        }
      }
  }
}
